calculator.py

```diff
@@ -0,0 +1,157 @@
+
+# Input criteria
+# -> one line / one input
+# -> start with 2 numbers 
+# -> 1 operation
+
+import time
+
+def instruction():
+    welcome_message = "Welcome to the calculator. \n----------\nPossible operations: \naddition \t-> + \nsubtraction \t-> - \nmultiplication \t-> * \ndivision \t-> / \n----------\nQUIT: q\n----------"
+    print(welcome_message)
+    return
+
+def convert(user_input):
+    invalid_number_warning = "Please enter numbers."
+    operator_warning = "You did not input a valid operator."
+    possible_operations = ["+", "*", "/", "-"]
+
+    x = 0 
+    y = 0
+    requested_operation = ""
+
+    for el in possible_operations: 
+        operator_index = user_input.find(el)
+
+        if operator_index == -1: 
+            continue
+
+        elif operator_index > 0: 
+            requested_operation = el
+
+            try: 
+                x = eval(user_input[0:operator_index])
+                y = eval(user_input[operator_index+1:])
+
+            except: 
+                print(invalid_number_warning)
+                x = None
+                y = None
+
+            break
+
+        elif operator_index == 0: 
+            requested_operation = el
+            if requested_operation == possible_operations[3]: 
+                operator_index = user_input[1:].find(el) + 1
+
+                try: 
+                    x = eval(user_input[0:operator_index])
+                    y = eval(user_input[operator_index+1:])
+
+                except: 
+                    print(invalid_number_warning)
+                    x = None
+                    y = None
+
+                break
+
+    if requested_operation == "": 
+        print(operator_warning)
+
+    return x, y, requested_operation
+
+def calculate(x, y, requested_operation): 
+    if requested_operation == "+":
+        return add(x, y)
+    elif requested_operation == "-":
+        return subtract(x, y)
+    elif requested_operation == "*":
+        return multiply(x, y)
+    elif requested_operation == "/":
+        return divide(x, y)
+
+def add(x, y): 
+    result = 0
+    result = x + y
+    return result
+
+def subtract(x, y): 
+    result = 0
+    result = x - y
+    return result
+
+def multiply(x, y): 
+    result = 0
+    result = x * y
+    return result
+
+def divide(x, y): 
+    result = 0
+    if y == 0: 
+        print("Division by 0.")
+        result = None
+    else: 
+        result = x / y
+    return result
+
+def calculator_main():
+    invalid_input_warning = "Cannot calculate, invalid input. Try again."
+    user_input = input("Calculate:").replace(" ", "")
+
+    if user_input == "Q" or user_input == "q":
+        return "quit_flag"
+
+    else: 
+
+        try: 
+            result = calculate(*convert(user_input))
+            print("RESULT: %.2f\n----------" %(result))
+            return result
+
+        except: 
+            print(invalid_input_warning)
+            return
```

Comment:
Hmm... you shouldn't need this exception handling here, since `convert` already has a similar handling.

The only possible error here is if `calculate` return `None` which will cause `print("RESULT: %.2F\n--------" %(None))` to blow up.

Maybe add an if statement to check before printing.

Reply:
thank you, done! Is an if/else statement preferable to try/except? Is that because it's a simpler operation?

Reply:
try/except is mostly useful where you anticipate possible exceptions. And in this case, there's no possibility of an exception being thrown here. Hence if/else is most suited for this.